substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoStripFeature.java

```diff
@@ -0,0 +1,120 @@
+import com.oracle.svm.core.BuildArtifacts;
+import com.oracle.svm.core.BuildArtifacts.ArtifactType;
+import com.oracle.svm.core.SubstrateOptions;
+import com.oracle.svm.core.feature.AutomaticallyRegisteredFeature;
+import com.oracle.svm.core.feature.InternalFeature;
+import com.oracle.svm.core.option.HostedOptionValues;
+import com.oracle.svm.core.option.SubstrateOptionsParser;
+import com.oracle.svm.core.util.InterruptImageBuilding;
+import com.oracle.svm.core.util.UserError;
+import com.oracle.svm.hosted.FeatureImpl.AfterImageWriteAccessImpl;
+import com.oracle.svm.hosted.c.util.FileUtils;
+
+@AutomaticallyRegisteredFeature
+public class NativeImageDebugInfoStripFeature implements InternalFeature {
+
+    @Override
+    public boolean isInConfiguration(IsInConfigurationAccess access) {
+        return SubstrateOptions.GenerateDebugInfo.getValue() > 0 && SubstrateOptions.StripDebugInfo.getValue();
+    }
+
+    @SuppressWarnings("try")
+    @Override
+    public void afterImageWrite(AfterImageWriteAccess access) {
+        AfterImageWriteAccessImpl accessImpl = (AfterImageWriteAccessImpl) access;
+        DebugContext debugContext = new Builder(HostedOptionValues.singleton(), new GraalDebugHandlersFactory(GraalAccess.getOriginalSnippetReflection())).build();
+        try (Indent indent = debugContext.logAndIndent("Stripping debuginfo")) {
+            switch (ObjectFile.getNativeFormat()) {
+                case ELF:
+                    stripLinux(accessImpl);
+                    break;
+                case MACH_O:
+                case PECOFF:
+                    if (SubstrateOptions.StripDebugInfo.hasBeenSet()) {
+                        System.out.println("Warning: Using " + SubstrateOptionsParser.commandArgument(SubstrateOptions.StripDebugInfo, "+") + " not supported on macOS and Windows");
+                    }
+                    break;
+                default:
+                    throw UserError.abort("Unsupported object file format");
+            }
+        }
+    }
+
+    private static void stripLinux(AfterImageWriteAccessImpl accessImpl) {
+        String objcopyExe = "objcopy";
+        String debugExtension = ".debug";
+        Path imagePath = accessImpl.getImagePath();
+        Path imageName = imagePath.getFileName();
+        Path outputDirectory = imagePath.getParent();
+        String debugInfoName = imageName + debugExtension;
+        boolean objcopyAvailable = false;
+        try {
+            objcopyAvailable = FileUtils.executeCommand(objcopyExe, "--version") == 0;
+        } catch (IOException e) {
+            /* Fall through to `if (!objcopyAvailable)` */
+        } catch (InterruptedException e) {
+            throw new InterruptImageBuilding("Interrupted during checking for " + objcopyExe + " availability");
+        }
+
+        if (!objcopyAvailable) {
+            System.out.printf("Warning: %s not available. Skip generation of debuginfo file %s%n", objcopyExe, debugInfoName);
```

Comment:
I find this message a bit ambiguous. It's not clear whether debuginfo won't be generated at all or not.

I suggest:
```suggestion
            System.out.printf("Warning: %s not available. Skipping generation of separate debuginfo file %s, debuginfo will remain embedded in the executable%n", objcopyExe, debugInfoName);
```